effectful/handlers/pyro.py

```diff
@@ -252,190 +265,264 @@ def __repr__(self):
         return f"Naming({self.name_to_dim})"
 
 
-class PositionalDistribution(pyro.distributions.torch_distribution.TorchDistribution):
-    """A distribution wrapper that lazily converts indexed dimensions to
-    positional.
-
-    """
-
-    indices: Mapping[Operation[[], int], int]
-
-    def __init__(
-        self, base_dist: pyro.distributions.torch_distribution.TorchDistribution
-    ):
-        self.base_dist = base_dist
-        self.indices = sizesof(base_dist)
+@defop
+def named_distribution(
+    d: Annotated[TorchDistribution, Scoped[A]],
+    *names: Annotated[Operation[[], int], Scoped[B]],
+) -> Annotated[TorchDistribution, Scoped[A | B]]:
+    match defterm(d):
+        case Term(op=_call, args=(dist_constr, *args)) if _call is call:
+            assert dist_constr is not dist.Independent
+            named_args = []
+            for a in args:
+                assert isinstance(a, torch.Tensor)
+                named_args.append(
+                    Indexable(typing.cast(torch.Tensor, a))[tuple(n() for n in names)]
+                )
+            assert callable(dist_constr)
+            return dist_constr(*named_args)
+        case _:
+            raise NotImplementedError
 
-        n_base = len(base_dist.batch_shape) + len(base_dist.event_shape)
-        self.naming = Naming.from_shape(self.indices.keys(), n_base)
 
-        super().__init__()
+@defop
+def positional_distribution(
```

Comment:
Hmm, why does `positional_distribution` needs to be a separate operation, as opposed to just using `to_tensor` on `defterm(d)`?

Reply:
We could move this logic into `to_tensor`, sure. The default behavior of `to_tensor` doesn't work though, because vmap is restricted to only return tensors.

Reply:
As discussed, I'd like to leave this alone for now.